File: py/ch02/E02A_motif_enumeration.py
# Implanted Motif Problem:
# Find all (k, d)-motifs in a collection of strings
#   Input: A collection of strings Dna, and integers k and d
#   Output: All (k, d)-motifs in Dna

from utils import text, read_lines
from ch01.E01N_d_neighborhood import Neighbors


def MotifEnumeration(dna, k, d):
    motifs = []

    for pattern in dna:
        temp = set()

        for i in range(len(pattern) - k + 1):
            neighbors = Neighbors(text(pattern, i, k), d)

            for word in neighbors:
                temp.add(word)

        for motif in temp:
            motifs.append(motif)

    motif_pattern = set()

    for element in motifs:
        if motifs.count(element) == len(dna):
            motif_pattern.add(element)

    return motif_pattern


if __name__ == "__main__":
    _dna, _k = read_lines(end_with='int')
    _d = int(input("d: "))

    for _word in MotifEnumeration(_dna, _k, _d):
        print(_word)

File: cpp/ch02/E02A_motif_enumeration.cpp
//
// Created by Matej Hrnčiar on 28/07/2022.
//

#include "E02A_motif_enumeration.h"

set<string> MotifEnumeration(const vector<string> &dna, int k, int d) {
    list<string> motifs;
    set<string> motif_pattern;

    for (auto &pattern : dna) {
        set<string> temp;

        for (int i = 0; i < pattern.size() - k + 1; i++) {
            auto neighbors = Neighbors(Text(pattern, i, k), d);

            for (const auto &word : neighbors) {
                temp.insert(word);
            }
        }

        for (auto &motif : temp) {
            motifs.push_back(motif);
        }
    }

    for (auto &element : motifs) {
        int total = 0;
        for (auto &temp : motifs) {
            if (element == temp)
                total++;
        }

        if (total == dna.size()) {
            motif_pattern.insert(element);
        }
    }

    return motif_pattern;
}

int main() {
    vector<string> _dna;
    string _temp;
    int _k, _d;

    cout << "k: ";
    cin >> _k;
    cout << "d: ";
    cin >> _d;

    _dna = ReadLines();

    auto _motifs = MotifEnumeration(_dna, _k, _d);

    for (const auto &word : _motifs) {
        cout << word << endl;
    }
}
